CRM/Core/Payment/Form.php

```diff
@@ -65,11 +65,15 @@ static public function setPaymentFieldsByProcessor(&$form, $processor, $forceBil
     $form->assign('paymentTypeName', $paymentTypeName);
     $form->assign('paymentTypeLabel', $paymentTypeLabel);
     $form->_paymentFields = $form->billingFieldSets[$paymentTypeName]['fields'] = self::getPaymentFieldMetadata($processor);
-    $form->_paymentFields = array_merge($form->_paymentFields, self::getBillingAddressMetadata($processor, $form->_bltID));
     $form->assign('paymentFields', self::getPaymentFields($processor));
-    self::setBillingAddressFields($form, $processor);
-    // @todo - this may be obsolete - although potentially it could be used to re-order things in the form.
-    $form->billingFieldSets['billing_name_address-group']['fields'] = array();
+    // return without adding billing fields if billing_mode = 4 (@todo - more the ability to set that to the payment processor)
+    // or payment processor is NULL (pay later)
+    if (CRM_Utils_Array::value('billing_mode', $processor) != 4) {
+      $form->_paymentFields = array_merge($form->_paymentFields, self::getBillingAddressMetadata($processor, $form->_bltID));
+      self::setBillingAddressFields($form, $processor);
```

Comment:
Ah - if I compare getBillingAddressMetadata with getPaymentFieldMetadata it looks like the latter is doing some filtering that the former isn't.

Now I think about it there is a reason for that - sometimes the fields are displayed on a different form (in non-core processors). I think they should both behave the same ie. not actually filter the returned fields, but not set them to required if they are not in the other array. 

I'm actually inclined to say NEVER set them to required here & leave that entirely to the validatePaymentInstrument function - I'm 80% to that point of view